source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc

```diff
@@ -127,19 +145,30 @@ void SimpleHttpCache::updateHeaders(const LookupContext& lookup_context,
     return;
   }
 
-  // https://www.rfc-editor.org/rfc/pdfrfc/rfc7234.txt.pdf
-  // 4.3.4 Freshening Stored Responses upon Validation
-  // use other header fields provided in the 304 (Not Modified)
-  // response to replace all instances of the corresponding header
-  // fields in the stored response.
-  //
   // Assumptions:
   // 1. The internet is fast, i.e. we get the result as soon as the server sends it.
-  // Race conditions would not be possible because we are always processing up-to-date data.
+  //    Race conditions would not be possible because we are always processing up-to-date data.
   // 2. No key collision for etag. Therefore, if etag matches it's the same resource.
   // 3. Backend is correct. etag is being used as a unique identifier to the resource
-  // TODO(tangsaidi) merge the header map instead of replacing it according to rfc7234
-  entry.response_headers_ = Http::createHeaderMap<Http::ResponseHeaderMapImpl>(response_headers);
+
+  // use other header fields provided in the new response to replace all instances
+  // of the corresponding header fields in the stored response
+  absl::flat_hash_set<Http::LowerCaseString> updatedHeaderFields;
+  response_headers.iterate(
+      [&entry, &updatedHeaderFields](
+          const Http::HeaderEntry& incoming_response_header) -> Http::HeaderMap::Iterate {
+        Http::LowerCaseString lower_case_key{incoming_response_header.key().getStringView()};
+        absl::string_view incoming_value{incoming_response_header.value().getStringView()};
+        if (headersNotToUpdate().contains(lower_case_key)) {
+          return Http::HeaderMap::Iterate::Continue;
+        }
+        if (!updatedHeaderFields.contains(lower_case_key)) {
```

Comment:
@jmarantz This `updatedHeaderFields` hash set handles the case where the incoming headers have multiple values for one field. A call to `setCopy` would overwrite previous values as we iterate through the response_headers. Another reason why I kept the `remove` and `addCopy` separate is because the behaviour of `setCopy` can be implementation dependent. In our current implementation, it is equivalent to `remove` then `add`. But [sometimes](https://github.com/envoyproxy/envoy/pull/9114/files#diff-2920212df144c38f64da08a4094a67b3c897669460307a6ed0147461fa2d4907R494), the `setCopy` would only replace the first occurrence.

Reply:
can you document that nuance in a comment here? Maybe also note that if setCopy is changed to remove all fields then this can be simplified.

Also as the fields in the set are well-known, does that multiple-value case ever occur for them?

Reply:
The link above directs to an old commit from 2 years ago. Our current `setCopy` implementation removes all fields, But we can't really simplify the code, because then we still need to use `addCopy` for special cases where the incoming headers have multiple-value. It's basically same amount of code
```   
if (updatedHeaderFields.contains(lower_case_key)) {
   addCopy(lower_case_key, value)
   updatedHeaderFields.insert(lower_case_key)
} else {
   setCopy(lower_case_key, value)
}
```
vs
```
if (!updatedHeaderFields.contains(lower_case_key)) {
   remove(lower_case_key)
   updatedHeaderFields.insert(lower_case_key)

}
   addCopy(lower_case_key, value)

```

And yes, the header_map may have multiple values for the same key because not all the headers are inlined.

Reply:
I'll add a comment to explain this